examples/compression/compression_middleware.go

```diff
@@ -0,0 +1,101 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/klauspost/compress/zstd"
+	"github.com/momentohq/client-sdk-go/config/middleware"
+	"github.com/momentohq/client-sdk-go/momento"
+	"github.com/momentohq/client-sdk-go/responses"
+)
+
+type compressionMiddleware struct {
+	middleware.Middleware
+	encoder *zstd.Encoder
+	decoder *zstd.Decoder
+}
+
+func (mw *compressionMiddleware) GetRequestHandler(baseHandler middleware.RequestHandler) (middleware.RequestHandler, error) {
+	return NewCompressionMiddlewareRequestHandler(baseHandler, mw.encoder, mw.decoder), nil
+}
+
+func NewCompressionMiddleware(props middleware.Props) middleware.Middleware {
+	// We use the default compression level for the encoder, but this could be made configurable.
+	encoder, _ := zstd.NewWriter(nil, zstd.WithEncoderLevel(zstd.SpeedDefault))
+	decoder, _ := zstd.NewReader(nil)
+	mw := middleware.NewMiddleware(props)
+	return &compressionMiddleware{mw, encoder, decoder}
+}
+
+type compressionMiddlewareRequestHandler struct {
+	middleware.RequestHandler
+	encoder *zstd.Encoder
+	decoder *zstd.Decoder
+}
+
+func NewCompressionMiddlewareRequestHandler(rh middleware.RequestHandler, encoder *zstd.Encoder, decoder *zstd.Decoder) middleware.RequestHandler {
+	return &compressionMiddlewareRequestHandler{rh, encoder, decoder}
+}
+
+func (rh *compressionMiddlewareRequestHandler) OnRequest(req interface{}) (interface{}, error) {
+	// Compress on writes
+	switch r := req.(type) {
+	case *momento.SetRequest:
+		rh.GetLogger().Info("Setting key: %s, value: %s", r.Key, r.Value)
+		rawData := r.Value
+		compressed := rh.encoder.EncodeAll([]byte(fmt.Sprintf("%v", rawData)), nil)
+		rh.GetLogger().Info("Compressed request %T: %d bytes -> %d bytes", req, len(fmt.Sprintf("%v", rawData)), len(compressed))
+		return &momento.SetRequest{
+			CacheName: r.CacheName,
+			Key:       r.Key,
+			Value:     momento.String(compressed),
+		}, nil
+	case *momento.SetIfAbsentOrHashEqualRequest:
+		rh.GetLogger().Info("Setting key: %s, value: %s", r.Key, r.Value)
+		rawData := r.Value
+		compressed := rh.encoder.EncodeAll([]byte(fmt.Sprintf("%v", rawData)), nil)
+		rh.GetLogger().Info("Compressed request %T: %d bytes -> %d bytes", req, len(fmt.Sprintf("%v", rawData)), len(compressed))
+		return &momento.SetIfAbsentOrHashEqualRequest{
+			CacheName: r.CacheName,
+			Key:       r.Key,
+			Value:     momento.String(compressed),
+			HashEqual: r.HashEqual,
+		}, nil
+	default:
+		rh.GetLogger().Info("[compressionMiddleware] No action for OnRequest type: %T", req)
```

Comment:
```suggestion
		rh.GetLogger().Info("No action for OnRequest type: %T", req)
```

The logger will take care of that.

Reply:
ah that's right, was leftover from testing with fmt.Printf statements